### PR TITLE
add KEEP_VOLUMES_NAMED feature

### PR DESCRIPTION
**Pull request by msutter**
This allows to pass a list of named volumes to keep.
Note that dangling unnamed volumes will always be removed.

### CODE REVIEW

**Comment by wzrdtales** [comment]
An important addition. @meltwater Is there anything missing to merge this or is this project not longer maintained?

File: README.md [hunk 1]
```diff
@@ -44,9 +45,7 @@ However it will not match
  * :tag
  * [[:alpha:]]:tag
 
-By default, both are set to **\*\*None\*\*** which is the same as the blank string.  If you want to keep ALL images or containers, effectively disabling this 
-part of the cleanup, then you should use **\*:\*** to match all images.  Do not 
-use a bare **\*** as this will be taken as a filename match.
+By default, both are set to \*\*None\*\* which is the same as the blank string.  If you want to keep ALL images, containers or volumes then you should use \*\*All\*\* to match all items. Do not use a bare **\*** as this will be taken as a filename match. Note that with KEEP_VOLUMES_NAMED, **All** will keep all named volumes, but still delete the unnamed ones.
```
Comment (jnjackins) [inline README.md:48]: You escaped the asterisks in the first `**All**`, but not the second.

File: run.sh [hunk 2]
```diff
@@ -76,27 +88,38 @@ do
...
     IFS='
  '
 
-    # Cleanup exited/dead containers
     echo "=> Removing exited/dead containers"
     EXITED_CONTAINERS_IDS="`docker ps -a -q -f status=exited -f status=dead | xargs echo`"
     for CONTAINER_ID in $EXITED_CONTAINERS_IDS; do
       CONTAINER_IMAGE=$(docker inspect --format='{{(index .Config.Image)}}' $CONTAINER_ID)
-      CONTAINER_NAME=$(docker inspect --format='{{(index .Name)}}' $CONTAINER_ID)
+      CONTAINER_NAME=$(docker inspect --format='{{(index .Name)}}' $CONTAINER_ID | sed 's/^\///')
```
Comment (jnjackins) [inline run.sh:122]: This change seems unrelated.

File: run.sh [hunk 3]
```diff
@@ -76,27 +88,38 @@ do
...
     else
       echo "=> Removing unused volumes using native 'docker volume' command"
-      for volume in $(docker volume ls -qf dangling=true); do
-        echo "Deleting ${volume}"
-        docker volume rm "${volume}"
+      DANGLING_VOLUMES_IDS="`docker volume ls -qf dangling=true | xargs echo`"
+      for VOLUME_ID in $DANGLING_VOLUMES_IDS; do
+        if [ $DEBUG ]; then echo "DEBUG: Check volume $VOLUME_ID"; fi
+        keepit=0
+        if [ ${#VOLUME_ID} -eq 64 ]; then
```
Comment (jnjackins) [inline run.sh:100]: `if echo $VOLUME_ID |grep -q "[0-9a-fA-F]\{64\}"; then`

File: run.sh [hunk 4]
```diff
@@ -76,27 +88,38 @@ do
...
-
     if [[ $(docker version --format '{{(index .Server.Version)}}' | grep -E '^[01]\.[012345678]\.') ]]; then
       echo "=> Removing unused volumes using 'docker-cleanup-volumes.sh' script"
       /docker-cleanup-volumes.sh
     else
       echo "=> Removing unused volumes using native 'docker volume' command"
-      for volume in $(docker volume ls -qf dangling=true); do
-        echo "Deleting ${volume}"
-        docker volume rm "${volume}"
+      DANGLING_VOLUMES_IDS="`docker volume ls -qf dangling=true | xargs echo`"
```
Comment (jnjackins) [inline run.sh:96]: What was the purpose of changing this from `for volume in $(docker volume ls -qf dangling=true); do`?

File: run.sh [hunk 5]
```diff
@@ -3,14 +3,17 @@
 checkPatterns() {
     keepit=$3
     if [ -n "$1" ]; then
-        for PATTERN in $(echo $1 | tr "," "\n"); do
-        if [[ "$2" = $PATTERN* ]]; then
-            if [ $DEBUG ]; then echo "DEBUG: Matches $PATTERN - keeping"; fi
-            keepit=1
+      for PATTERN in $(echo $1 | tr "," "\n"); do
+        if [[ $PATTERN = '.' ]]; then
```
Comment (jnjackins) [inline run.sh:7]: I don't understand how (or if) the old code here worked. Are you fixing a bug or adding new functionality?

**Review by jnjackins — CHANGES_REQUESTED** [review]
I apologize that it has take so much time to give this attention. Please review my comments.

**Comment by jnjackins** [comment]
Fixes #32

**Comment by rasjoh** [comment]
Won't fix due to retiring project

**Pull request closed without merging**